File: Code/Cpp/DetectPlates.cpp
// DetectPlates.cpp
//  _     _                           ____  _       _         ____                            _ _   _
// | |   (_) ___ ___ _ __  ___  ___  |  _ \| | __ _| |_ ___  |  _ \ ___  ___ ___   __ _ _ __ (_) |_(_) ___  _ __
// | |   | |/ __/ _ \ '_ \/ __|/ _ \ | |_) | |/ _` | __/ _ \ | |_) / _ \/ __/ _ \ / _` | '_ \| | __| |/ _ \| '_ \
// | |___| | (__  __/ | | \__ \  __/ |  __/| | (_| | |_  __/ |  _ <  __/ (__ (_) | (_| | | | | | |_| | (_) | | | |
// |_____|_|\___\___|_| |_|___/\___| |_|   |_|\__,_|\__\___| |_| \_\___|\___\___/ \__, |_| |_|_|\__|_|\___/|_| |_|
//                                                                                |___/
// (c) Shahar Gino, July-2017, sgino209@gmail.com

#include "DetectPlates.hpp"

// ------------------------------------------------------------------------------------------------------------------------------
std::vector<PossiblePlate> detectPlatesInScene(cv::Mat &imgOriginalScene, plate_params_t plate_params) {
    
    cv::RNG rng;
    cv::Mat imgThreshScene;
    cv::Mat imgGrayscaleScene;
    std::vector<PossiblePlate> vectorOfPossiblePlates;
    char buffer[100];

    // Pre-processing (CSC --> contrast --> blur --> threshold):
    preprocess(imgOriginalScene,
               imgGrayscaleScene,
               imgThreshScene,
               plate_params.PreprocessGaussKernel,
               plate_params.PreprocessThreshBlockSize,
               plate_params.PreprocessThreshweight,
               plate_params.PreprocessMorphKernel,
               plate_params.OpMode, "platesDet");

    // Find all possible characters in the scene (finds all contours that could be characters, w/o OCR yet):
    std::vector<PossibleChar> vectorOfPossibleCharsInScene = findPossibleCharsInImage(imgThreshScene,
                                                                                      plate_params.MinPixelWidth,  plate_params.MaxPixelWidth,
                                                                                      plate_params.MinPixelHeight, plate_params.MaxPixelHeight,
                                                                                      plate_params.MinAspectRatio, plate_params.MaxAspectRatio,
                                                                                      plate_params.MinPixelArea,   plate_params.MaxPixelArea,
                                                                                      plate_params.debugMode);
    
    // Given a list of all possible chars, find groups of matching characters (later on, each group will attempt to be recognized as a plate):
    std::vector<std::vector<PossibleChar> > vectorOfVectorsOfMatchingCharsInScene = findVectorOfVectorsOfMatchingChars(vectorOfPossibleCharsInScene,
                                                                                                                       plate_params.MinNumberOfMatchingChars,
                                                                                                                       plate_params.MaxNumberOfMatchingChars,
                                                                                                                       plate_params.MinAngleBetweenChars,
                                                                                                                       plate_params.MaxAngleBetweenChars,
                                                                                                                       plate_params.MinChangeInArea,
                                                                                                                       plate_params.MaxChangeInArea,
                                                                                                                       plate_params.MinChangeInWidth,
                                                                                                                       plate_params.MaxChangeInWidth,
                                                                                                                       plate_params.MinChangeInHeight,
                                                                                                                       plate_params.MaxChangeInHeight,
                                                                                                                       plate_params.MaxDiagSizeMultipleAway);
    
    // For each group of matching chars, attempt to extract plate:
    for (auto &vectorOfMatchingChars : vectorOfVectorsOfMatchingCharsInScene) {
        
        PossiblePlate possiblePlate = extractPlate(imgOriginalScene,
                                                   vectorOfMatchingChars,
                                                   plate_params.PlateWidthPaddingFactor,
                                                   plate_params.PlateHeightPaddingFactor);
        
        // Add plate to list of possible plates (if found):
        if (!possiblePlate.imgPlate.empty()) {

            // Verify that the plate has enough saturation:
            cv::Mat imgHSV, imgS;
            std::vector<cv::Mat> vectorOfBGRImages, vectorOfHSVImages;
            cv::cvtColor(possiblePlate.imgPlate, imgHSV, CV_BGR2HSV);
            cv::split(imgHSV, vectorOfHSVImages);
            imgS = vectorOfHSVImages[1];
            int N = imgS.total();
            int histSize = 256;
            float range[] = { 0, 256 };
            const float* histRange = { range };
            cv::MatND s_hist;
            cv::calcHist( &imgS, 1, 0, cv::Mat(), s_hist, 1, &histSize, &histRange, true, false );
            float hist_norm = 0;
            for( int h = 0; h < histSize; h++ ) { 
                hist_norm += h * s_hist.at<float>(h)/N; 
            }

            if (hist_norm > plate_params.MinHistNormThr) {
                vectorOfPossiblePlates.push_back(possiblePlate);
            } 
            else if (plate_params.debugMode) {

                char buffer[100];
                sprintf(buffer, "Plates rejected: HistNorm=%.2f", hist_norm);
                debug(buffer);
            }
        }
    }
    
    // Add rectangle plate candidate:
    if (plate_params.FindRectangledPlate) {
    
        PossiblePlate possiblePlate = findRectangledPlate(imgOriginalScene,
                                                          imgThreshScene,
                                                          0.45,   // circularity_min
                                                          0.65,   // circularity_max
                                                          3.5,    // aspect_ratio_min
                                                          4.4,    // aspect_ratio_max
                                                          0.005,  // area_norm_min
                                                          0.03,   // area_norm_max
                                                          plate_params.debugMode);
    
        if (!possiblePlate.imgPlate.empty()) {
            possiblePlate.rectFind = true;
            vectorOfPossiblePlates.push_back(possiblePlate);
        }
    }

    if (plate_params.debugMode) {
      sprintf(buffer, "%lu possible plates found", vectorOfPossiblePlates.size());
      debug(buffer);
    }

    // -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- ..
    if (plate_params.debugMode) {
    
        // Original image:
        cv::imwrite("img_original.jpg", imgOriginalScene);
        
        // Pre-processing images:
        cv::imwrite("img_gray.jpg",      imgGrayscaleScene);
        cv::imwrite("img_threshold.jpg", imgThreshScene);
        
        // Possible characters in image:
        cv::Mat imgContours(imgOriginalScene.size(), CV_8UC3, SCALAR_BLACK);
        std::vector<std::vector<cv::Point> > contours;
        
        for (auto &possibleChar : vectorOfPossibleCharsInScene) {
            contours.push_back(possibleChar.contour);
        }
        cv::drawContours(imgContours, contours, -1, SCALAR_WHITE);
        cv::imwrite("img_contours_possible_chars.jpg", imgContours);
        
        // Matching characters:
        sprintf(buffer, "#listOfListsOfMatchingCharsInScene = %lu", vectorOfVectorsOfMatchingCharsInScene.size());
        debug(buffer);
        
        imgContours = cv::Mat(imgOriginalScene.size(), CV_8UC3, SCALAR_BLACK);
        
        for (auto &vectorOfMatchingChars : vectorOfVectorsOfMatchingCharsInScene) {
            int intRandomBlue = rng.uniform(0, 256);
            int intRandomGreen = rng.uniform(0, 256);
            int intRandomRed = rng.uniform(0, 256);
            
            std::vector<std::vector<cv::Point> > contours;
            
            sprintf(buffer, "#listOfMatchingChars = %lu", vectorOfMatchingChars.size());
            debug(buffer);
            for (auto &matchingChar : vectorOfMatchingChars) {
                contours.push_back(matchingChar.contour);
            }
            cv::drawContours(imgContours, contours, -1, cv::Scalar((double)intRandomBlue, (double)intRandomGreen, (double)intRandomRed));
        }
        cv::imwrite("img_contours_matching_chars.jpg", imgContours);
        
        // Possible license-plates:
        for (unsigned int i = 0; i < vectorOfPossiblePlates.size(); i++) {
            cv::Point2f p2fRectPoints[4];
            
            vectorOfPossiblePlates[i].rrLocationOfPlateInScene.points(p2fRectPoints);
            
            for (int j = 0; j < 4; j++) {
                const cv::Scalar color = vectorOfPossiblePlates[i].rectFind ? SCALAR_GREEN : SCALAR_RED;
                cv::line(imgContours, p2fRectPoints[j], p2fRectPoints[(j + 1) % 4], color, 2);
            }
            char buffer [50];
            sprintf (buffer, "img_contours_possible_plates_%d.jpg", i);
            cv::imwrite(buffer, imgContours);
            sprintf (buffer, "img_plate_%d.jpg", i);
            cv::imwrite(buffer, vectorOfPossiblePlates[i].imgPlate);
        }
        
        debug("Plate detection complete");
    }

    return vectorOfPossiblePlates;
}

// ------------------------------------------------------------------------------------------------------------------------------
PossiblePlate extractPlate(cv::Mat &imgOriginal, std::vector<PossibleChar> &vectorOfMatchingChars, double PlateWidthPaddingFactor, double PlateHeightPaddingFactor) {
    
    PossiblePlate possiblePlate;

    // Sort characters from left to right based on x position:
    std::sort(vectorOfMatchingChars.begin(), vectorOfMatchingChars.end(), PossibleChar::sortCharsLeftToRight);

    // Calculate the plate centroid (average of leftmost and righhtmost characters):
    double dblPlateCenterX = (double)(vectorOfMatchingChars[0].intCenterX + vectorOfMatchingChars[vectorOfMatchingChars.size() - 1].intCenterX) / 2.0;
    double dblPlateCenterY = (double)(vectorOfMatchingChars[0].intCenterY + vectorOfMatchingChars[vectorOfMatchingChars.size() - 1].intCenterY) / 2.0;
    cv::Point2d p2dPlateCenter(dblPlateCenterX, dblPlateCenterY);

    // Calculate plate width (rightmost - leftmost characters):
    int intPlateWidth = (int)(PlateWidthPaddingFactor * (vectorOfMatchingChars[vectorOfMatchingChars.size() - 1].boundingRect.x +
                                                         vectorOfMatchingChars[vectorOfMatchingChars.size() - 1].boundingRect.width -
                                                         vectorOfMatchingChars[0].boundingRect.x));

    // Calculate plate height (average over all characters):
    double intTotalOfCharHeights = 0;
    for (auto &matchingChar : vectorOfMatchingChars) {
        intTotalOfCharHeights = intTotalOfCharHeights + matchingChar.boundingRect.height;
    }
    double dblAverageCharHeight = (double)intTotalOfCharHeights / vectorOfMatchingChars.size();
    int intPlateHeight = (int)(dblAverageCharHeight * PlateHeightPaddingFactor);

    // Calculate correction angle of plate region (simple geometry calculation):
    double dblOpposite = vectorOfMatchingChars[vectorOfMatchingChars.size() - 1].intCenterY - vectorOfMatchingChars[0].intCenterY;
    double dblHypotenuse = distanceBetweenChars(vectorOfMatchingChars[0], vectorOfMatchingChars[vectorOfMatchingChars.size() - 1]);
    double dblCorrectionAngleInRad = asin(dblOpposite / dblHypotenuse);
    double dblCorrectionAngleInDeg = dblCorrectionAngleInRad * (180.0 / CV_PI);

    // Rotate the entire image (affine warp), for compensating the angle of the plate region:
    cv::Mat imgRotated;
    cv::Mat rotationMatrix;
    rotationMatrix = cv::getRotationMatrix2D(p2dPlateCenter, dblCorrectionAngleInDeg, 1.0);
    cv::warpAffine(imgOriginal, imgRotated, rotationMatrix, imgOriginal.size());

    // Crop the plate from the image:
    cv::Mat imgCropped;
    possiblePlate.rrLocationOfPlateInScene = cv::RotatedRect(p2dPlateCenter, cv::Size2f((float)intPlateWidth, (float)intPlateHeight), (float)dblCorrectionAngleInDeg);
    cv::getRectSubPix(imgRotated, possiblePlate.rrLocationOfPlateInScene.size, possiblePlate.rrLocationOfPlateInScene.center, imgCropped);

    // Create and return possiblePlate object, which packs most the above information:
    possiblePlate.imgPlate = imgCropped;

    return(possiblePlate);
}

// ------------------------------------------------------------------------------------------------------------------------------
PossiblePlate findRectangledPlate(cv::Mat &imgOriginalScene, cv::Mat &imgThresh, double circularity_min, double circularity_max,
                                  double aspect_ratio_min, double aspect_ratio_max, double area_norm_min, double area_norm_max, bool debugMode) {
    
    // Initialization:
    cv::Rect res_box;
    int rect_cntr = 0;
    int max_rect_area = 0;
    double res_area_norm = 0;
    double res_circularity = 0;
    double res_aspect_ratio = 0;
    PossiblePlate possiblePlate;
    cv::Point2d p2dPlateCenter(0,0);
    
    // Find contours:
    std::vector<std::vector<cv::Point> > contours;
    cv::findContours(imgThresh.clone(), contours, CV_RETR_LIST, CV_CHAIN_APPROX_NONE);

    // For each contour:
    for (std::vector<cv::Point>& contour : contours) {
        
        // Compute convex hull:
        std::vector<cv::Point> hull;
        convexHull(contour, hull);
        
        // Compute circularity, used for shape classification:
        double area = contourArea(hull);
        double area_norm = contourArea(hull) / imgThresh.size().area();
        double perimeter = arcLength(hull, true);
        double circularity = (4 * CV_PI * area) / (perimeter * perimeter);
        
        // Shape classification (rectangle):
        if ((circularity > circularity_min) && (circularity < circularity_max) && (area > max_rect_area)) {
        
            cv::Rect box = boundingRect(contour);
            double aspect_ratio = (double)box.width / box.height;
            
            if ((aspect_ratio > aspect_ratio_min) && (aspect_ratio < aspect_ratio_max) &&
                (area_norm > area_norm_min) && (area_norm < area_norm_max)) {
                
                // Registration:
                res_box = box;
                max_rect_area = area;
                res_area_norm = area_norm;
                res_aspect_ratio = aspect_ratio;
                res_circularity = circularity;

                for ( size_t i=0; i<contour.size(); i++ )
                {
                    p2dPlateCenter.x += contour[i].x;
                    p2dPlateCenter.y += contour[i].y;
                }
                p2dPlateCenter.x /= contour.size();
                p2dPlateCenter.y /= contour.size();
                
                rect_cntr++;
            }
        }
    }
    
    if (rect_cntr > 0) {

        // Crop the plate from the image:
        cv::Mat imgCropped;
        possiblePlate.rrLocationOfPlateInScene = cv::RotatedRect(p2dPlateCenter, cv::Size2f((float)res_box.width, (float)res_box.height), 0);
        cv::getRectSubPix(imgOriginalScene, possiblePlate.rrLocationOfPlateInScene.size, possiblePlate.rrLocationOfPlateInScene.center, imgCropped);
        possiblePlate.imgPlate = imgCropped;
        
        // -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- ..
        if (debugMode) {
            
            char buffer[200];
            sprintf(buffer, "rect #%d: box.w=%d, box.h=%d, AR=%.4f, area=%.4f, circ=%.4f",
                    rect_cntr-1, res_box.width, res_box.height, res_aspect_ratio, res_area_norm, res_circularity);
            debug(buffer);
        }
    }
    
    return possiblePlate;
}

File: Code/Python/DetectPlates.py
#!/usr/bin/python
#  _     _                           ____  _       _         ____                            _ _   _
# | |   (_) ___ ___ _ __  ___  ___  |  _ \| | __ _| |_ ___  |  _ \ ___  ___ ___   __ _ _ __ (_) |_(_) ___  _ __
# | |   | |/ __/ _ \ '_ \/ __|/ _ \ | |_) | |/ _` | __/ _ \ | |_) / _ \/ __/ _ \ / _` | '_ \| | __| |/ _ \| '_ \
# | |___| | (__  __/ | | \__ \  __/ |  __/| | (_| | |_  __/ |  _ <  __/ (__ (_) | (_| | | | | | |_| | (_) | | | |
# |_____|_|\___\___|_| |_|___/\___| |_|   |_|\__,_|\__\___| |_| \_\___|\___\___/ \__, |_| |_|_|\__|_|\___/|_| |_|
#                                                                                |___/
# (c) Shahar Gino, July-2017, sgino209@gmail.com

from math import asin, pi
from random import randint
from Preprocess import preprocess
from PossiblePlate import PossiblePlate
from Auxiliary import info, debug, Colors
from numpy import zeros, uint8
from DetectChars import findListOfListsOfMatchingChars, findPossibleCharsInImage
from cv2 import line, boxPoints, warpAffine, drawContours, findContours, getRectSubPix, convexHull, arcLength, split, \
    getRotationMatrix2D, imwrite, RETR_LIST, CHAIN_APPROX_NONE, contourArea, boundingRect, cvtColor, calcHist, COLOR_BGR2HSV

# ---------------------------------------------------------------------------------------------------------------
def detectPlatesInScene(imgOriginalScene,
                        PreprocessGaussKernel, PreprocessThreshBlockSize, PreprocessThreshweight, PreprocessMorphKernel,
                        PlateWidthPaddingFactor, PlateHeightPaddingFactor, FindRectangledPlate,
                        MinPixelWidth, MaxPixelWidth, MinPixelHeight, MaxPixelHeight, MinAspectRatio, MaxAspectRatio, MinPixelArea, MaxPixelArea,
                        MaxDiagSizeMultipleAway, MinNumberOfMatchingChars, MaxNumberOfMatchingChars, MinAngleBetweenChars, MaxAngleBetweenChars,
                        MinChangeInArea, MaxChangeInArea, MinChangeInWidth, MaxChangeInWidth, MinChangeInHeight, MaxChangeInHeight, MinHistNormThr, OpMode, debugMode):
    """ License Plate Detection in a given input image scene, using geometrical analysis techniques """

    # Pre-processing (CSC --> contrast --> blur --> threshold):
    imgGrayscaleScene, imgThreshScene = preprocess(imgOriginalScene,
                                                   PreprocessGaussKernel,
                                                   PreprocessThreshBlockSize,
                                                   PreprocessThreshweight,
                                                   PreprocessMorphKernel,
                                                   OpMode, "platesDet")

    # Find all possible characters in the scene (finds all contours that could be characters, w/o OCR yet):
    listOfPossibleCharsInScene = findPossibleCharsInImage(imgThreshScene,
                                                          MinPixelWidth, MaxPixelWidth,
                                                          MinPixelHeight, MaxPixelHeight,
                                                          MinAspectRatio, MaxAspectRatio,
                                                          MinPixelArea, MaxPixelArea,
                                                          debugMode)

    # Given a list of all possible chars, find groups of matching characters (later on, each group will attempt to be recognized as a plate):
    listOfListsOfMatchingCharsInScene = findListOfListsOfMatchingChars(listOfPossibleCharsInScene,
                                                                       MinNumberOfMatchingChars,
                                                                       MaxNumberOfMatchingChars,
                                                                       MinAngleBetweenChars, MaxAngleBetweenChars,
                                                                       MinChangeInArea, MaxChangeInArea,
                                                                       MinChangeInWidth, MaxChangeInWidth,
                                                                       MinChangeInHeight, MaxChangeInHeight,
                                                                       MaxDiagSizeMultipleAway)

    # For each group of matching chars, attempt to extract plate:
    listOfPossiblePlates = []
    for listOfMatchingChars in listOfListsOfMatchingCharsInScene:

        possiblePlate = extractPlate(imgOriginalScene,
                                     listOfMatchingChars,
                                     PlateWidthPaddingFactor,
                                     PlateHeightPaddingFactor)

        # Add plate to list of possible plates (if found):
        if possiblePlate.imgPlate is not None:
            
            imgHSV = cvtColor(possiblePlate.imgPlate, COLOR_BGR2HSV)
            _, imgS, _ = split(imgHSV)
            N = imgS.size
            histSize = 256
            s_hist = calcHist([imgS], [0], None, [histSize], [0,256])
            hist_norm = 0
            for h in range(histSize):
              hist_norm += h * s_hist[h][0]/N

            # Verify that the plate has enough saturation:
            if hist_norm > MinHistNormThr:
                listOfPossiblePlates.append(possiblePlate)
            else:
                debug("Plates rejected: HistNorm=%.2f" % hist_norm, debugMode)

    # Add rectangle plate candidate:
    if (FindRectangledPlate):

        possiblePlate = findRectangledPlate(imgOriginalScene,
                                            imgThreshScene,
                                            0.45,   # circularity_min
                                            0.65,   # circularity_max
                                            3.5,    # aspect_ratio_min
                                            4.4,    # aspect_ratio_max
                                            0.005,  # area_norm_min
                                            0.03,   # area_norm_max
                                            debugMode)

        if (possiblePlate.imgPlate is not None):
            possiblePlate.rectFind = True
            listOfPossiblePlates.append(possiblePlate)

    debug("%d possible plates found" % len(listOfPossiblePlates), debugMode)

    # -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- .. -- ..
    if debugMode:

        height, width, _ = imgOriginalScene.shape

        # Original image:
        imwrite("img_original.jpg", imgOriginalScene)

        # Pre-processing images:
        imwrite("img_gray.jpg", imgGrayscaleScene)
        imwrite("img_threshold.jpg", imgThreshScene)

        # Possible characters in image:
        imgContours = zeros((height, width, 3), uint8)
        contours = []
        for possibleChar in listOfPossibleCharsInScene:
            contours.append(possibleChar.contour)
        drawContours(imgContours, contours, -1, Colors.white)
        imwrite("img_contours_possible_chars.jpg", imgContours)

        # Matching characters:
        debug("#listOfListsOfMatchingCharsInScene = %d" % len(listOfListsOfMatchingCharsInScene), True)
        imgContours = zeros((height, width, 3), uint8)
        for listOfMatchingChars in listOfListsOfMatchingCharsInScene:
            intRandomBlue = randint(0, 255)
            intRandomGreen = randint(0, 255)
            intRandomRed = randint(0, 255)
            contours = []
            debug("#listOfMatchingChars = %d" % len(listOfMatchingChars), True)
            for matchingChar in listOfMatchingChars:
                contours.append(matchingChar.contour)
            drawContours(imgContours, contours, -1, (intRandomBlue, intRandomGreen, intRandomRed))
            imwrite("img_contours_matching_chars.jpg", imgContours)

        # Possible license-plates:
        for i in range(0, len(listOfPossiblePlates)):
            p2fRectPoints = boxPoints(listOfPossiblePlates[i].rrLocationOfPlateInScene)
            color = Colors.red
            if listOfPossiblePlates[i].rectFind:
                color = Colors.green
            line(imgContours, tuple(p2fRectPoints[0]), tuple(p2fRectPoints[1]), color, 2)
            line(imgContours, tuple(p2fRectPoints[1]), tuple(p2fRectPoints[2]), color, 2)
            line(imgContours, tuple(p2fRectPoints[2]), tuple(p2fRectPoints[3]), color, 2)
            line(imgContours, tuple(p2fRectPoints[3]), tuple(p2fRectPoints[0]), color, 2)
            imwrite("img_contours_possible_plates_%d.jpg" % i, imgContours)
            imwrite("img_plate_%d.jpg" % i, listOfPossiblePlates[i].imgPlate)

        debug("Plate detection complete", True)

    return listOfPossiblePlates

# ---------------------------------------------------------------------------------------------------------------
def extractPlate(imgOriginal, listOfMatchingChars, PlateWidthPaddingFactor, PlateHeightPaddingFactor):
    """ Extract license-plate in the provided image, based on given contours group that corresponds for matching characters """

    # Sort characters from left to right based on x position:
    listOfMatchingChars.sort(key=lambda matchingChar_: matchingChar_.intCenterX)

    # Calculate the plate centroid (average of leftmost and righhtmost characters):
    fltPlateCenterX = (listOfMatchingChars[0].intCenterX + listOfMatchingChars[len(listOfMatchingChars) - 1].intCenterX) / 2.0
    fltPlateCenterY = (listOfMatchingChars[0].intCenterY + listOfMatchingChars[len(listOfMatchingChars) - 1].intCenterY) / 2.0
    ptPlateCenter = fltPlateCenterX, fltPlateCenterY

    # Calculate plate width (rightmost - leftmost characters):
    intPlateWidth = int(PlateWidthPaddingFactor * (listOfMatchingChars[len(listOfMatchingChars) - 1].intBoundingRectX +
                                                   listOfMatchingChars[len(listOfMatchingChars) - 1].intBoundingRectWidth -
                                                   listOfMatchingChars[0].intBoundingRectX))

    # Calculate plate height (average over all characters):
    intTotalOfCharHeights = 0
    for matchingChar in listOfMatchingChars:
        intTotalOfCharHeights = intTotalOfCharHeights + matchingChar.intBoundingRectHeight
    fltAverageCharHeight = intTotalOfCharHeights / len(listOfMatchingChars)
    intPlateHeight = int(fltAverageCharHeight * PlateHeightPaddingFactor)

    # Calculate correction angle of plate region (simple geometry calculation):
    fltOpposite = listOfMatchingChars[len(listOfMatchingChars) - 1].intCenterY - listOfMatchingChars[0].intCenterY
    fltHypotenuse = (listOfMatchingChars[0] - listOfMatchingChars[len(listOfMatchingChars) - 1])
    fltCorrectionAngleInRad = asin(fltOpposite / fltHypotenuse)
    fltCorrectionAngleInDeg = fltCorrectionAngleInRad * (180.0 / pi)

    # Rotate the entire image (affine warp), for compensating the angle of the plate region:
    rotationMatrix = getRotationMatrix2D(tuple(ptPlateCenter), fltCorrectionAngleInDeg, 1.0)
    height, width, _ = imgOriginal.shape
    imgRotated = warpAffine(imgOriginal, rotationMatrix, (width, height))

    # Crop the plate from the image:
    imgCropped = getRectSubPix(imgRotated, (intPlateWidth, intPlateHeight), tuple(ptPlateCenter))

    # Create and return possiblePlate object, which packs most the above information:
    possiblePlate = PossiblePlate()
    possiblePlate.rrLocationOfPlateInScene = (tuple(ptPlateCenter), (intPlateWidth, intPlateHeight), fltCorrectionAngleInDeg)
    possiblePlate.imgPlate = imgCropped

    return possiblePlate

# ---------------------------------------------------------------------------------------------------------------
def findRectangledPlate(imgOriginal, imgThresh, circularity_min, circularity_max, aspect_ratio_min,
                        aspect_ratio_max, area_norm_min, area_norm_max, debugMode):
    """ Find rectangles """

    # Initialization:
    rect_cntr = 0
    max_rect_area = 0
    res_area_norm = 0
    res_circularity = 0
    res_aspect_ratio = 0
    fltPlateCenterX = 0
    fltPlateCenterY = 0
    possiblePlate = PossiblePlate()

    # Find all contours in plate:
    imgThreshCopy = imgThresh.copy()
    _, contours, _ = findContours(imgThreshCopy, RETR_LIST, CHAIN_APPROX_NONE)

    # For each contour, check if it might be representing a rectangled plate:
    for contour in contours:

        # Compute convex hull:
        hull = convexHull(contour)

        # Compute circularity, used for shape classification:
        area = contourArea(hull)
        area_norm = contourArea(hull) / imgThresh.size
        perimeter = arcLength(hull, True)
        circularity = 0
        if perimeter > 0:
            circularity = (4 * pi * area) / (perimeter * perimeter)

        # Shape classification (rectangle):
        if ((circularity > circularity_min) and (circularity < circularity_max) and (area > max_rect_area)):

            box_x, box_y, box_width, box_height = boundingRect(contour)
            aspect_ratio = box_width / box_height

            if ((aspect_ratio > aspect_ratio_min) and (aspect_ratio < aspect_ratio_max) and
                (area_norm > area_norm_min) and (area_norm < area_norm_max)):

                # Registration:
                res_box = [box_x, box_y, box_width, box_height]
                max_rect_area = area
                res_area_norm = area_norm
                res_aspect_ratio = aspect_ratio
                res_circularity = circularity

                for i in range(0, len(contour)):

                    fltPlateCenterX += contour[i][0][0]
                    fltPlateCenterY += contour[i][0][1]

                fltPlateCenterX /= len(contour)
                fltPlateCenterY /= len(contour)

                rect_cntr += 1

    # Return result:
    if rect_cntr > 0:

        ptPlateCenter = fltPlateCenterX, fltPlateCenterY

        # Crop the plate from the image:
        possiblePlate.rrLocationOfPlateInScene = (tuple(ptPlateCenter), (res_box[2], res_box[3]), 0)
        imgCropped = getRectSubPix(imgOriginal, possiblePlate.rrLocationOfPlateInScene[1], possiblePlate.rrLocationOfPlateInScene[0])
        possiblePlate.imgPlate = imgCropped

        debug("rect #%d: box.w=%d, box.h=%d, AR=%.4f, area=%.4f, circ=%.4f" %
              (rect_cntr-1, res_box[2], res_box[3], res_aspect_ratio, res_area_norm, res_circularity), debugMode)

    return possiblePlate
